File: numerics/src/mixture_hmm_em.py
"""EM for univariate HMMs with Gaussian-mixture emissions.

The hidden state process is a standard finite-state Markov chain. Conditional
on hidden state k, the observation density is a finite mixture of M Gaussian
components. This is used as an enriched-emission diagnostic: if a corrective
extra Gaussian HMM state disappears after allowing mixtures within aggregate
states, the extra state is more naturally interpreted as an emission correction.
"""
from __future__ import annotations

from dataclasses import dataclass, field

import numpy as np

from .hmm_em import SIGMA_MIN, _fb_core


@dataclass
class MixtureHMMFit:
    """Best EM fit over random starts, states sorted by emission mean."""
    K: int
    M: int
    loglik: float
    pi: np.ndarray              # (K,)
    A: np.ndarray               # (K, K)
    weights: np.ndarray         # (K, M)
    means: np.ndarray           # (K, M)
    sds: np.ndarray             # (K, M)
    gamma: np.ndarray           # (T, K)
    component_resp: np.ndarray  # (T, K, M)
    viterbi: np.ndarray         # (T,)
    n_iter: int
    loglik_trace: list = field(default_factory=list)

    def state_means(self) -> np.ndarray:
        return np.sum(self.weights * self.means, axis=1)

    def state_vars(self) -> np.ndarray:
        second = np.sum(self.weights * (self.sds ** 2 + self.means ** 2), axis=1)
        return second - self.state_means() ** 2

    def n_params(self) -> int:
        transition = (self.K - 1) + self.K * (self.K - 1)
        emissions = self.K * ((self.M - 1) + 2 * self.M)
        return transition + emissions

    def bic(self, T: int) -> float:
        return -2.0 * self.loglik + self.n_params() * np.log(T)


def _norm_pdf(y: np.ndarray, means: np.ndarray, sds: np.ndarray) -> np.ndarray:
    var = sds ** 2
    out = np.exp(-0.5 * (y[:, None, None] - means[None, :, :]) ** 2 / var[None, :, :])
    out /= np.sqrt(2.0 * np.pi * var[None, :, :])
    return np.maximum(out, 1e-300)


def _emission_probs(y, weights, means, sds):
    comp = _norm_pdf(y, means, sds)
    B = np.sum(weights[None, :, :] * comp, axis=2)
    return np.maximum(B, 1e-300), comp


def forward_backward_mixture(y, pi, A, weights, means, sds):
    """Scaled forward-backward for mixture-emission HMMs."""
    B, comp = _emission_probs(y, weights, means, sds)
    loglik, gamma, xi_sum = _fb_core(B, pi, A)
    numer = gamma[:, :, None] * weights[None, :, :] * comp
    component_resp = numer / np.maximum(B[:, :, None], 1e-300)
    return float(loglik), gamma, xi_sum, component_resp


def viterbi_path_mixture(y, pi, A, weights, means, sds):
    """Viterbi decoding with state-level mixture emission densities."""
    T, K = len(y), len(pi)
    B, _ = _emission_probs(y, weights, means, sds)
    logB = np.log(B)
    logA = np.log(np.maximum(A, 1e-300))
    delta = np.zeros((T, K))
    psi = np.zeros((T, K), dtype=int)
    delta[0] = np.log(np.maximum(pi, 1e-300)) + logB[0]
    for t in range(1, T):
        cand = delta[t - 1][:, None] + logA
        psi[t] = cand.argmax(axis=0)
        delta[t] = cand.max(axis=0) + logB[t]
    path = np.zeros(T, dtype=int)
    path[-1] = int(delta[-1].argmax())
    for t in range(T - 2, -1, -1):
        path[t] = psi[t + 1, path[t + 1]]
    return path


def _initial_components(y, K, M, rng, sigma_min):
    y_sd = max(float(np.std(y)), sigma_min)
    q_state = np.linspace(0.15, 0.85, K)
    state_centers = np.quantile(y, q_state) + rng.normal(0.0, 0.15 * y_sd, K)
    offsets = np.linspace(-0.35, 0.35, M) * y_sd
    means = state_centers[:, None] + offsets[None, :] + rng.normal(0.0, 0.08 * y_sd, (K, M))
    sds = np.full((K, M), max(2 * sigma_min, 0.45 * y_sd)) * rng.uniform(0.7, 1.3, (K, M))
    weights = np.vstack([rng.dirichlet(np.ones(M) + 1.0) for _ in range(K)])
    pi = rng.dirichlet(np.ones(K))
    A = np.vstack([rng.dirichlet(np.ones(K) + 6.0 * np.eye(K)[i]) for i in range(K)])
    return pi, A, weights, means, sds


def _sort_components(weights, means, sds):
    weights = weights.copy()
    means = means.copy()
    sds = sds.copy()
    for k in range(weights.shape[0]):
        order = np.argsort(means[k])
        weights[k] = weights[k, order]
        means[k] = means[k, order]
        sds[k] = sds[k, order]
    return weights, means, sds


def _em_loop(y, pi, A, weights, means, sds, max_iter, tol, sigma_min):
    """Run mixture-emission EM from supplied parameters."""
    trace = []
    prev = -np.inf
    for it in range(max_iter):
        loglik, gamma, xi_sum, comp_resp = forward_backward_mixture(
            y, pi, A, weights, means, sds)
        trace.append(loglik)
        if loglik < prev - 1e-6 * (1.0 + abs(prev)):
            raise RuntimeError(
                f"Mixture EM log-likelihood decreased at iteration {it}: {prev} -> {loglik}"
            )
        converged = (loglik - prev) < tol * (1.0 + abs(prev))
        prev = loglik

        pi = gamma[0] + 1e-10
        pi /= pi.sum()
        A = xi_sum + 1e-10
        A /= A.sum(axis=1, keepdims=True)

        state_weights = gamma.sum(axis=0) + 1e-12
        comp_counts = comp_resp.sum(axis=0) + 1e-12
        weights = comp_counts / state_weights[:, None]
        weights /= weights.sum(axis=1, keepdims=True)
        means = np.sum(comp_resp * y[:, None, None], axis=0) / comp_counts
        var = np.sum(comp_resp * (y[:, None, None] - means[None, :, :]) ** 2, axis=0)
        var /= comp_counts
        sds = np.sqrt(np.maximum(var, sigma_min ** 2))
        weights, means, sds = _sort_components(weights, means, sds)
        if converged and it > 0:
            break

    loglik, gamma, _, comp_resp = forward_backward_mixture(y, pi, A, weights, means, sds)
    trace.append(loglik)
    return dict(loglik=loglik, pi=pi, A=A, weights=weights, means=means, sds=sds,
                gamma=gamma, component_resp=comp_resp, n_iter=len(trace) - 1,
                trace=trace)


def _em_single(y, K, M, rng, max_iter, tol, sigma_min):
    pi, A, weights, means, sds = _initial_components(y, K, M, rng, sigma_min)
    return _em_loop(y, pi, A, weights, means, sds, max_iter, tol, sigma_min)


def fit_mixture_hmm(y, K, M=2, seed=1, n_starts=12, max_iter=250, tol=1e-6,
                    sigma_min=SIGMA_MIN, screen_iter=None, refine_top=None,
                    screen_tol=1e-4) -> MixtureHMMFit:
    """Fit a K-state HMM with M Gaussian components per state."""
    rng = np.random.default_rng(seed)
    best = None
    if screen_iter is not None and refine_top is not None and refine_top < n_starts:
        screened = [
            _em_single(y, K, M, rng, int(screen_iter), screen_tol, sigma_min)
            for _ in range(n_starts)
        ]
        selected = sorted(screened, key=lambda res: res["loglik"], reverse=True)[
            :int(refine_top)
        ]
        for res in selected:
            refined = _em_loop(y, res["pi"], res["A"], res["weights"],
                               res["means"], res["sds"], max_iter, tol,
                               sigma_min)
            if best is None or refined["loglik"] > best["loglik"]:
                best = refined
    else:
        for _ in range(n_starts):
            res = _em_single(y, K, M, rng, max_iter, tol, sigma_min)
            if best is None or res["loglik"] > best["loglik"]:
                best = res

    state_mean = np.sum(best["weights"] * best["means"], axis=1)
    order = np.argsort(state_mean)
    pi = best["pi"][order]
    A = best["A"][np.ix_(order, order)]
    weights = best["weights"][order]
    means = best["means"][order]
    sds = best["sds"][order]
    gamma = best["gamma"][:, order]
    comp_resp = best["component_resp"][:, order, :]
    vit = viterbi_path_mixture(y, pi, A, weights, means, sds)
    return MixtureHMMFit(K=K, M=M, loglik=best["loglik"], pi=pi, A=A,
                         weights=weights, means=means, sds=sds, gamma=gamma,
                         component_resp=comp_resp, viterbi=vit,
                         n_iter=best["n_iter"], loglik_trace=best["trace"])


def validate_mixture_fit(fit: MixtureHMMFit, T: int) -> None:
    """Assertions on a fitted mixture-emission HMM."""
    assert fit.gamma.shape == (T, fit.K), "gamma has wrong shape"
    assert fit.component_resp.shape == (T, fit.K, fit.M), "component_resp has wrong shape"
    assert np.allclose(fit.A.sum(axis=1), 1.0, atol=1e-8), "rows of A must sum to 1"
    assert np.allclose(fit.pi.sum(), 1.0, atol=1e-8), "pi must sum to 1"
    assert np.allclose(fit.weights.sum(axis=1), 1.0, atol=1e-8), \
        "mixture weights must sum to 1 within state"
    assert np.all(fit.sds > 0), "estimated sds must be positive"
    assert np.allclose(fit.gamma.sum(axis=1), 1.0, atol=1e-8), \
        "posterior probabilities must sum to 1 at each time"
    assert np.all(np.isfinite(fit.gamma)), "gamma contains NaN/Inf"
    assert np.all(np.isfinite(fit.component_resp)), "component_resp contains NaN/Inf"
    assert np.all(np.isfinite([fit.loglik, *fit.means.ravel(), *fit.sds.ravel()])), \
        "parameters contain NaN/Inf"
    tr = np.array(fit.loglik_trace)
    assert np.all(np.diff(tr) >= -1e-6 * (1.0 + np.abs(tr[:-1]))), \
        "mixture EM log-likelihood is not non-decreasing (beyond tolerance)"

File: numerics/R/mixture_hmm_em.R
# Expectation-maximization for univariate HMMs with Gaussian-mixture emissions.
# Source numerics/R/hmm_em.R before using this file.

component_mask <- function(component_counts, K, M) {
  component_counts <- as.integer(component_counts)
  if (length(component_counts) != K || anyNA(component_counts) ||
      any(component_counts < 1L) || any(component_counts > M)) {
    stop("component_counts must have K entries between 1 and M.", call. = FALSE)
  }
  outer(component_counts, seq_len(M), FUN = `>=`)
}

mixture_emission_probabilities <- function(y, weights, means, sds) {
  T <- length(y)
  K <- nrow(weights)
  M <- ncol(weights)
  component <- array(0, dim = c(T, K, M))
  for (state in seq_len(K)) {
    for (component_index in seq_len(M)) {
      component[, state, component_index] <- pmax(
        stats::dnorm(
          y,
          mean = means[state, component_index],
          sd = sds[state, component_index]
        ),
        1e-300
      )
    }
  }
  emission <- matrix(0, nrow = T, ncol = K)
  for (state in seq_len(K)) {
    component_matrix <- matrix(
      component[, state, ], nrow = T, ncol = M
    )
    emission[, state] <- component_matrix %*% weights[state, ]
  }
  list(emission = pmax(emission, 1e-300), component = component)
}

forward_backward_mixture <- function(
    y, initial, transition, weights, means, sds, lengths = NULL) {
  emission_result <- mixture_emission_probabilities(y, weights, means, sds)
  emission <- emission_result$emission
  K <- length(initial)
  M <- ncol(weights)
  posterior <- matrix(0, nrow = length(y), ncol = K)
  transition_sum <- matrix(0, nrow = K, ncol = K)
  log_likelihood <- 0

  for (indices in sequence_indices(length(y), lengths)) {
    result <- forward_backward_core(emission[indices, , drop = FALSE], initial, transition)
    posterior[indices, ] <- result$posterior
    transition_sum <- transition_sum + result$transition_sum
    log_likelihood <- log_likelihood + result$log_likelihood
  }

  component_posterior <- array(0, dim = c(length(y), K, M))
  for (state in seq_len(K)) {
    for (component_index in seq_len(M)) {
      component_posterior[, state, component_index] <-
        posterior[, state] * weights[state, component_index] *
        emission_result$component[, state, component_index] /
        emission[, state]
    }
  }

  list(
    log_likelihood = log_likelihood,
    posterior = posterior,
    transition_sum = transition_sum,
    component_posterior = component_posterior
  )
}

viterbi_single_mixture <- function(y, initial, transition, weights, means, sds) {
  T <- length(y)
  K <- length(initial)
  log_emission <- log(
    mixture_emission_probabilities(y, weights, means, sds)$emission
  )
  log_transition <- log(pmax(transition, 1e-300))
  delta <- matrix(0, nrow = T, ncol = K)
  back_pointer <- matrix(1L, nrow = T, ncol = K)
  delta[1L, ] <- log(pmax(initial, 1e-300)) + log_emission[1L, ]

  if (T > 1L) {
    for (time in 2:T) {
      for (destination in seq_len(K)) {
        candidates <- delta[time - 1L, ] + log_transition[, destination]
        back_pointer[time, destination] <- which.max(candidates)
        delta[time, destination] <- max(candidates) + log_emission[time, destination]
      }
    }
  }

  path <- integer(T)
  path[[T]] <- which.max(delta[T, ])
  if (T > 1L) {
    for (time in seq.int(T - 1L, 1L)) {
      path[[time]] <- back_pointer[time + 1L, path[[time + 1L]]]
    }
  }
  path
}

viterbi_mixture <- function(
    y, initial, transition, weights, means, sds, lengths = NULL) {
  path <- integer(length(y))
  for (indices in sequence_indices(length(y), lengths)) {
    path[indices] <- viterbi_single_mixture(
      y[indices], initial, transition, weights, means, sds
    )
  }
  path
}

initial_mixture_parameters <- function(y, K, M, component_counts, sigma_min) {
  y_sd <- max(stats::sd(y), sigma_min)
  state_centers <- as.numeric(
    stats::quantile(y, seq(0.15, 0.85, length.out = K), names = FALSE)
  ) + stats::rnorm(K, 0, 0.15 * y_sd)
  means <- matrix(rep(state_centers, M), nrow = K, ncol = M)
  sds <- matrix(max(2 * sigma_min, 0.45 * y_sd), nrow = K, ncol = M)
  weights <- matrix(0, nrow = K, ncol = M)

  for (state in seq_len(K)) {
    count <- component_counts[[state]]
    offsets <- if (count > 1L) seq(-0.35, 0.35, length.out = count) * y_sd else 0
    means[state, seq_len(count)] <- state_centers[[state]] + offsets +
      stats::rnorm(count, 0, 0.08 * y_sd)
    sds[state, seq_len(count)] <- sds[state, seq_len(count)] *
      stats::runif(count, 0.7, 1.3)
    weights[state, seq_len(count)] <- rdirichlet1(rep(2, count))
  }

  initial <- rdirichlet1(rep(1, K))
  transition <- t(vapply(
    seq_len(K),
    function(state) rdirichlet1(rep(1, K) + 6 * (seq_len(K) == state)),
    numeric(K)
  ))
  list(
    initial = initial,
    transition = transition,
    weights = weights,
    means = means,
    sds = sds
  )
}

sort_mixture_components <- function(weights, means, sds, component_counts) {
  for (state in seq_len(nrow(weights))) {
    count <- component_counts[[state]]
    ordering <- order(means[state, seq_len(count)])
    weights[state, seq_len(count)] <- weights[state, ordering]
    means[state, seq_len(count)] <- means[state, ordering]
    sds[state, seq_len(count)] <- sds[state, ordering]
  }
  list(weights = weights, means = means, sds = sds)
}

mixture_em_loop <- function(
    y, initial, transition, weights, means, sds, component_counts,
    max_iter, tolerance, sigma_min, lengths = NULL) {
  trace <- numeric()
  previous <- -Inf
  K <- length(initial)
  M <- ncol(weights)
  active <- component_mask(component_counts, K, M)
  sequences <- sequence_indices(length(y), lengths)
  starts <- vapply(sequences, function(index) index[[1L]], integer(1))

  for (iteration in seq_len(max_iter)) {
    fb <- forward_backward_mixture(
      y, initial, transition, weights, means, sds, lengths
    )
    log_likelihood <- fb$log_likelihood
    trace <- c(trace, log_likelihood)
    if (is.finite(previous) &&
        log_likelihood < previous - 1e-6 * (1 + abs(previous))) {
      stop(
        sprintf(
          "Mixture EM log-likelihood decreased at iteration %d: %.12f to %.12f.",
          iteration, previous, log_likelihood
        ),
        call. = FALSE
      )
    }
    converged <- is.finite(previous) &&
      (log_likelihood - previous) < tolerance * (1 + abs(previous))
    previous <- log_likelihood

    posterior <- fb$posterior
    initial <- normalize_probability(colMeans(posterior[starts, , drop = FALSE]) + 1e-10)
    transition <- normalize_rows(fb$transition_sum + 1e-10)

    component_totals <- apply(fb$component_posterior, c(2L, 3L), sum)
    weights <- ifelse(active, component_totals + 1e-12, 0)
    weights <- weights / rowSums(weights)
    safe_totals <- ifelse(active, component_totals + 1e-12, 1)

    updated_means <- matrix(0, nrow = K, ncol = M)
    updated_variances <- matrix(0, nrow = K, ncol = M)
    for (state in seq_len(K)) {
      for (component_index in seq_len(M)) {
        responsibility <- fb$component_posterior[, state, component_index]
        updated_means[state, component_index] <-
          sum(responsibility * y) / safe_totals[state, component_index]
      }
    }
    means[active] <- updated_means[active]
    for (state in seq_len(K)) {
      for (component_index in seq_len(M)) {
        responsibility <- fb$component_posterior[, state, component_index]
        updated_variances[state, component_index] <- sum(
          responsibility * (y - means[state, component_index])^2
        ) / safe_totals[state, component_index]
      }
    }
    sds[active] <- sqrt(pmax(updated_variances[active], sigma_min^2))

    sorted <- sort_mixture_components(weights, means, sds, component_counts)
    weights <- sorted$weights
    means <- sorted$means
    sds <- sorted$sds

    if (converged && iteration > 1L) {
      break
    }
  }

  fb <- forward_backward_mixture(
    y, initial, transition, weights, means, sds, lengths
  )
  trace <- c(trace, fb$log_likelihood)
  list(
    log_likelihood = fb$log_likelihood,
    initial = initial,
    transition = transition,
    weights = weights,
    means = means,
    sds = sds,
    posterior = fb$posterior,
    component_posterior = fb$component_posterior,
    n_iter = length(trace) - 1L,
    log_likelihood_trace = trace
  )
}

mixture_em_start <- function(
    y, K, M, component_counts, max_iter, tolerance, sigma_min,
    lengths = NULL) {
  initial_values <- initial_mixture_parameters(
    y, K, M, component_counts, sigma_min
  )
  mixture_em_loop(
    y,
    initial_values$initial,
    initial_values$transition,
    initial_values$weights,
    initial_values$means,
    initial_values$sds,
    component_counts,
    max_iter,
    tolerance,
    sigma_min,
    lengths
  )
}

mixture_n_parameters <- function(fit) {
  transition <- (fit$K - 1L) + fit$K * (fit$K - 1L)
  emissions <- sum((fit$component_counts - 1L) + 2L * fit$component_counts)
  as.integer(transition + emissions)
}

mixture_state_means <- function(fit) {
  rowSums(fit$weights * fit$means)
}

mixture_state_variances <- function(fit) {
  second <- rowSums(fit$weights * (fit$sds^2 + fit$means^2))
  second - mixture_state_means(fit)^2
}

fit_mixture_hmm <- function(
    y, K, M = 2L, seed = 1L, n_starts = 12L, max_iter = 250L,
    tolerance = 1e-6, sigma_min = SIGMA_MIN, screen_iter = NULL,
    refine_top = NULL, screen_tolerance = 1e-4,
    component_counts = rep(M, K), lengths = NULL) {
  K <- as.integer(K)
  M <- as.integer(M)
  component_counts <- as.integer(component_counts)
  component_mask(component_counts, K, M)
  set.seed(seed)

  if (!is.null(screen_iter) && !is.null(refine_top) && refine_top < n_starts) {
    screened <- lapply(seq_len(n_starts), function(unused) {
      mixture_em_start(
        y, K, M, component_counts, as.integer(screen_iter),
        screen_tolerance, sigma_min, lengths
      )
    })
    selected <- screened[order(
      vapply(screened, `[[`, numeric(1), "log_likelihood"),
      decreasing = TRUE
    )[seq_len(refine_top)]]
    candidates <- lapply(selected, function(result) {
      mixture_em_loop(
        y, result$initial, result$transition, result$weights,
        result$means, result$sds, component_counts, max_iter, tolerance,
        sigma_min, lengths
      )
    })
  } else {
    candidates <- lapply(seq_len(n_starts), function(unused) {
      mixture_em_start(
        y, K, M, component_counts, max_iter, tolerance, sigma_min, lengths
      )
    })
  }

  best <- candidates[[which.max(vapply(
    candidates, `[[`, numeric(1), "log_likelihood"
  ))]]
  state_means <- rowSums(best$weights * best$means)
  ordering <- order(state_means)
  fit <- list(
    model = "Mixture-emission HMM",
    K = K,
    M = M,
    component_counts = component_counts[ordering],
    log_likelihood = best$log_likelihood,
    initial = best$initial[ordering],
    transition = best$transition[ordering, ordering, drop = FALSE],
    weights = best$weights[ordering, , drop = FALSE],
    means = best$means[ordering, , drop = FALSE],
    sds = best$sds[ordering, , drop = FALSE],
    posterior = best$posterior[, ordering, drop = FALSE],
    component_posterior = best$component_posterior[, ordering, , drop = FALSE],
    n_iter = best$n_iter,
    log_likelihood_trace = best$log_likelihood_trace
  )
  fit$n_parameters <- mixture_n_parameters(fit)
  fit$viterbi <- viterbi_mixture(
    y, fit$initial, fit$transition, fit$weights, fit$means, fit$sds, lengths
  )
  class(fit) <- c("mixture_hmm_fit", "hmm_fit")
  fit
}

validate_mixture_fit <- function(fit, T) {
  active <- component_mask(fit$component_counts, fit$K, fit$M)
  stopifnot(
    inherits(fit, "mixture_hmm_fit"),
    identical(dim(fit$posterior), c(as.integer(T), fit$K)),
    identical(dim(fit$component_posterior), c(as.integer(T), fit$K, fit$M)),
    max(abs(rowSums(fit$transition) - 1)) < 1e-8,
    abs(sum(fit$initial) - 1) < 1e-8,
    max(abs(rowSums(fit$weights) - 1)) < 1e-8,
    all(abs(fit$weights[!active]) < 1e-12),
    all(fit$sds > 0),
    max(abs(rowSums(fit$posterior) - 1)) < 1e-8,
    all(is.finite(fit$posterior)),
    all(is.finite(fit$component_posterior)),
    all(is.finite(c(fit$log_likelihood, fit$means, fit$sds)))
  )
  differences <- diff(fit$log_likelihood_trace)
  tolerance <- -1e-6 * (1 + abs(head(fit$log_likelihood_trace, -1L)))
  stopifnot(all(differences >= tolerance))
  invisible(TRUE)
}
